src/config/graph.rs

```diff
@@ -0,0 +1,476 @@
+                    Some(Node::Transform { .. }) => "transform",
+                    Some(Node::Sink { .. }) => "sink",
+                    _ => panic!("only transforms and sinks have inputs"),
+                };
+                errors.push(format!(
+                    "Input \"{}\" for {} \"{}\" doesn't match any components.",
+                    edge.from, output_type, edge.to
+                ));
+            }
+        }
+
+        if errors.is_empty() {
+            Ok(())
+        } else {
+            Err(errors)
+        }
+    }
+}
+
+impl From<&ConfigBuilder> for Graph {
+    fn from(config: &ConfigBuilder) -> Self {
+        let mut graph = Graph::default();
+
+        for (id, config) in config.sources.iter() {
+            graph.add_source(id.clone(), config.inner.output_type());
+        }
+
+        for (id, config) in config.transforms.iter() {
+            graph.add_transform(
+                id.clone(),
+                config.inner.input_type(),
+                config.inner.output_type(),
+                config.inputs.clone(),
+            );
+            for name in config.inner.named_outputs() {
+                graph.add_transform_output(id, name.to_string());
+            }
+        }
+
+        for (id, config) in config.sinks.iter() {
+            graph.add_sink(id.clone(), config.inner.input_type(), config.inputs.clone());
+        }
+
+        graph
+    }
+}
+
+fn paths_rec(
+    graph: &Graph,
+    node: &ComponentKey,
+    mut path: Vec<ComponentKey>,
+) -> Result<Vec<Vec<ComponentKey>>, String> {
+    if let Some(i) = path.iter().position(|p| p == node) {
+        let mut segment = path.split_off(i);
+        segment.push(node.into());
+        // I think this is maybe easier to grok from source -> sink, but I'm not
+        // married to either.
+        segment.reverse();
+        return Err(format!(
+            "Cyclic dependency detected in the chain [ {} ]",
+            segment
+                .iter()
+                .map(|item| item.to_string())
+                .collect::<Vec<_>>()
+                .join(" -> ")
+        ));
+    }
+    path.push(node.clone());
+    match graph.nodes.get(node) {
+        Some(Node::Source { .. }) | None => {
+            path.reverse();
+            Ok(vec![path])
+        }
+        Some(Node::Transform { .. }) | Some(Node::Sink { .. }) => {
+            let inputs = graph
+                .edges
+                .iter()
+                .filter(|e| &e.to == node)
+                .map(|e| e.from.clone());
+            let mut paths = Vec::new();
+            for input in inputs {
+                match paths_rec(graph, &input, path.clone()) {
+                    Ok(mut p) => paths.append(&mut p),
+                    Err(err) => {
+                        return Err(err);
+                    }
+                }
+            }
+            Ok(paths)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
It would be valuable, considering how central this will be to vector's good working order, to push for property tests for the graph code. I'm happy to help with that.